xdsl/transforms/arith_add_fastmath.py

```diff
@@ -0,0 +1,90 @@
+"""Passes to manipulate fastmath flags in FP arith operations."""
+
+from dataclasses import dataclass
+from typing import Literal
+
+from xdsl.dialects import arith, builtin, llvm
+from xdsl.ir import Operation
+from xdsl.passes import MLContext, ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+)
+
+
+def _match_string_to_enum(flag: str, enum_type: type[llvm.FastMathFlag]):
+    for member in enum_type:
+        if flag == str(member.value):
+            return member
+
+    raise ValueError(f"{flag} is not a valid fastmath flag.")
+
+
+def _get_flag_list(flags: list[str], enum_type: type[llvm.FastMathFlag]):
+    list_flags: list[enum_type] = []
+    for flag in flags:
+        list_flags.append(_match_string_to_enum(flag, llvm.FastMathFlag))
+    return list_flags
+
+
+@dataclass
+class AddArithFastMathFlags(RewritePattern):
+    """Adds fastmath flags to FP binary operations from arith dialect."""
+
+    arith_op_cls: type[arith.FloatingPointLikeBinaryOp]
+    fastmath_op_attr: arith.FastMathFlagsAttr
+
+    def match_and_rewrite(self, op: Operation, rewriter: PatternRewriter) -> None:
+        if not isinstance(op, self.arith_op_cls):
+            return
+
+        op.fastmath = self.fastmath_op_attr
```

Comment:
Wouldn't
```suggestion
    @op_type_rewrite_pattern
    def match_and_rewrite(self, op: arith.Addf | arith.Subf | arith.Mulf | arith.Divf | arith.Minimumf | arith.Maximumf, rewriter: PatternRewriter) -> None:
        op.fastmath = self.fastmath_op_attr
```
achieve what you hardcode at the Pass level either way? Now, maybe I'm missing a rationale to not use it.

Reply:
Yeah, agreed, not sure what's best.

We do what I'm doing at the pass level at the `arith` lowering to `riscv` [here](https://github.com/xdslproject/xdsl/blob/b9b6223fac1528b94116057891cfc1f7289b28cb/xdsl/backend/riscv/lowering/convert_arith_to_riscv.py#L172).

Not sure what are the pros and cons for each, since we need to spell out explicitly the related operations at some point.

Reply:
Interesting! I'd say it makes more sense in that lowering case, where you map each type to another. So having the decorator do the check to then map the actual type would be paying overhead, where here I don't think you ever want to do more than skip non-listed types?

But anyway, I guess if we want to be perfectionists about it, the right way is a Trait/Interface: MLIR defines `ArithFastMathInterface` for this! (Used in math et al. too)

So I'd say let's just merge this as a simple helper, and see when we want a properly generic one :slightly_smiling_face:

Reply:
What decorator overhead? I'm not using one.